cpmpy/transformations/linearize.py

```diff
@@ -146,26 +152,69 @@ def linearize_constraint(lst_of_expr, supported={"sum","wsum"}, reified=False):
+                        #   e.g., 7 % -5 = 2 and -7 % 5 = -2
+                        # implement x % y == z as k * y + z == x with |z| < |y| and sign(x) = sign(z)
+                        # https://marcelkliemannel.com/articles/2021/dont-confuse-integer-division-with-floor-division/
                         x, y = lhs.args
                         lby, uby = get_bounds(y)
                         if lby <= 0 <= uby:
-                            raise NotImplementedError("Modulo with a divisor domain containing 0 is not supported. "
-                                                      "Please safen the expression first.")
-                        k = intvar(*get_bounds((x - rhs) // y))
-                        mult_res, newcons = get_or_make_var(k * y)
-                        remainder = rhs < abs(y)
-                        if is_true_cst(remainder):
-                            remainder = BoolVal(True)
-                        elif is_false_cst(remainder):
-                            remainder = BoolVal(False)
-                        newcons.append(remainder)
-                        newlist += linearize_constraint(newcons, supported, reified=reified)
+                            raise ValueError("Attempting linerarization of unsafe modulo, safen expression first (cpmpy/transformations/safen.py)")
 
+                        # k * y + z == x
+                        k = intvar(*get_bounds((x - rhs) // y))
+                        mult_res, side_cons = get_or_make_var(k * y)
                         cpm_expr = (mult_res + rhs) == x
+                        # |z| < |y|
+                        abs_of_z = cp.intvar(*get_bounds(abs(rhs)))
+                        if "abs" in supported:
+                            side_cons.append(abs(rhs) == abs_of_z)
+                        else: # need to linearize abs...
+                            z_is_pos = cp.boolvar()
+                            side_cons += [z_is_pos.implies(rhs >= 0), (~z_is_pos).implies(rhs < 0)]
+                            side_cons += [z_is_pos.implies(abs_of_z == rhs), (~z_is_pos).implies(abs_of_z == -rhs)]
+                        # TODO: do the following in constructor of abs instead?
+                        if lby >= 0:
+                            side_cons.append(abs_of_z < y)
+                        if uby <= 0:
+                            side_cons.append(abs_of_z < -y)
+                        # sign(x) = sign(z)
+                        lbx,ubx = get_bounds(x)
+                        if lbx >= 0:
+                            side_cons.append(rhs >= 0)
+                        elif ubx <= 0:
+                            side_cons.append(rhs <= 0)
+                        else: # x can be pos or neg
+                            x_is_pos, x_is_neg = cp.boolvar(), cp.boolvar()
+                            side_cons += [(~x_is_pos).implies(x <= 0), (~x_is_neg).implies(x >= 0),
```

Comment:
I think this doesn't totally define x_is_pos and x_is_neg. Can't we accomplish the same with only 1 boolvar and totally defining it?
_(~x_is_pos).implies(x <= 0),
(x_is_pos).implies(x >= 0),
x_is_pos.implies(rhs >= 0), 
(~x_is_pos).implies(rhs < 0)_

Reply:
Aha yes indeed! Adapted it.